File: Leetcode/MinimumCosttoConvertStringI.rs
// https://leetcode.com/problems/minimum-cost-to-convert-string-i

use std::collections::VecDeque;

struct Solution {}

impl Solution {
    pub fn minimum_cost(
        source: String,
        target: String,
        original: Vec<char>,
        changed: Vec<char>,
        cost: Vec<i32>,
    ) -> i64 {
        let original = original
            .into_iter()
            .map(|x| (x as u8 - b'a') as usize)
            .collect::<Vec<_>>();
        let changed = changed
            .into_iter()
            .map(|x| (x as u8 - b'a') as usize)
            .collect::<Vec<_>>();
        let cost = cost.into_iter().map(|x| x as i64).collect::<Vec<_>>();

        let mut mtx = vec![vec![i64::MAX; 26]; 26];
        for i in 0..26 {
            mtx[i][i] = 0;
        }

        for i in 0..original.len() {
            let cur = original[i];
            let mut cost_sum = 0;

            let mut q = VecDeque::<usize>::new();
            q.push_back(cur);
            while let Some(cur) = q.pop_front() {
                for j in 0..original.len() {
                    if original[j] == cur {
                        let to = changed[j];
                        if mtx[original[i]][to] > mtx[original[i]][cur] + cost[j] {
                            mtx[original[i]][to] = mtx[original[i]][cur] + cost[j];
                            q.push_back(to);
                        }
                    }
                }
            }
        }

        let source = source.into_bytes();
        let target = target.into_bytes();

        let mut ans = 0;
        for i in 0..source.len() {
            let from = (source[i] - b'a') as usize;
            let to = (target[i] - b'a') as usize;

            if mtx[from][to] == i64::MAX {
                ans = -1;
                break;
            } else {
                ans += mtx[from][to];
            }
        }

        ans
    }
}

File: Leetcode/MinimumCosttoConvertStringI.cpp
// https://leetcode.com/problems/minimum-cost-to-convert-string-i

class Solution {
   public:
    long long minimumCost(string source,
                          string target,
                          vector<char>& original,
                          vector<char>& changed,
                          vector<int>& cost) {
        const int NS = 26;
        const auto INF = numeric_limits<int>::max();

        vector<vector<int>> mtx(NS, vector<int>(NS, INF));

        for (int i = 0; i < original.size(); ++i) {
            const auto u = original[i] - 'a';
            const auto v = changed[i] - 'a';

            mtx[u][v] = min(mtx[u][v], cost[i]);
        }

        for (int o = 0; o < NS; ++o) {
            for (int u = 0; u < NS; ++u) {
                for (int v = 0; v < NS; ++v) {
                    for (int k = 0; k < NS; ++k) {
                        auto& uv = mtx[u][v];
                        const auto& uk = mtx[u][k];
                        const auto& kv = mtx[k][v];

                        if (uk != INF && kv != INF && uk + kv < uv) {
                            uv = uk + kv;
                        }
                    }
                }
            }
        }

        long long ans = 0;
        for (int i = 0; i < source.size(); ++i) {
            const auto u = source[i] - 'a';
            const auto v = target[i] - 'a';

            if (u == v) {
                continue;
            }

            if (mtx[u][v] == INF) {
                return -1;
            }

            ans += mtx[u][v];
        }

        return ans;
    }
};